alf/algorithms/particle_vi_algorithm.py

```diff
@@ -0,0 +1,298 @@
+            :math:`\nabla_x K(x, y)` (Tensor): the derivative of RBF kernel of shape (Nx x Ny x D)
+            
+        """
+        Nx, Dx = x.shape
+        Ny, Dy = y.shape
+        assert Dx == Dy
+        diff = x.unsqueeze(1) - y.unsqueeze(0)  # [Nx, Ny, W]
+        dist_sq = torch.sum(diff**2, -1)  # [Nx, Ny]
+        # h = self._kernel_width(dist_sq.view(-1))
+        h, _ = torch.median(dist_sq.view(-1), dim=0)
+        if h == 0.:
+            h = torch.ones_like(h)
+        else:
+            h = h / max(np.log(Nx), 1.)
+
+        kappa = torch.exp(-dist_sq / h)  # [Nx, Nx]
+        kappa_grad = torch.einsum('ij,ijk->ijk', kappa,
+                                  -2 * diff / h)  # [Nx, Ny, W]
+        return kappa, kappa_grad
+
+    def _score_func(self, x, alpha=1e-5):
+        r"""
+        Compute the stein estimator of the score function 
+        :math:`\nabla\log q = -(K + \alpha I)^{-1}\nabla K`,
+        used by gfsf_grad. 
+
+        Args:
+            x (Tensor): set of N particles, shape (N x D), where D is the 
+                dimenseion of each particle
+            alpha (float): weight of regularization for inverse kernel
+                this parameter turns out to be crucial for convergence.
+
+        Returns:
+            :math:`\nabla\log q` (Tensor): the score function of shape (N x D)
+            
+        """
+        N, D = x.shape
+        diff = x.unsqueeze(1) - x.unsqueeze(0)  # [N, N, D]
+        dist_sq = torch.sum(diff**2, -1)  # [N, N]
+        h, _ = torch.median(dist_sq.view(-1), dim=0)
+        h = h / np.log(N)
+
+        kappa = torch.exp(-dist_sq / h)  # [N, N]
+        kappa_inv = torch.inverse(kappa + alpha * torch.eye(N))  # [N, N]
+        kappa_grad = torch.einsum('ij,ijk->jk', kappa, -2 * diff / h)  # [N, D]
+
+        return kappa_inv @ kappa_grad
+
+    def _svgd_grad(self,
+                   particles,
+                   loss_func,
+                   entropy_regularization,
+                   transform_func=None):
+        """
+        Compute particle gradients via SVGD, empirical expectation
+        evaluated by splitting half of the sampled batch. 
+        """
+        if transform_func is not None:
+            particles, extra_particles = transform_func(particles)
+            aug_particles = torch.cat([particles, extra_particles], dim=-1)
+        else:
+            aug_particles = particles
+        loss_inputs = particles
+        loss = loss_func(loss_inputs)
+        if isinstance(loss, tuple):
+            neglogp = loss.loss
+        else:
+            neglogp = loss
+        loss_grad = torch.autograd.grad(neglogp.sum(),
+                                        loss_inputs)[0]  # [N, D]
+
+        # [N, N], [N, N, D]
+        kernel_weight, kernel_grad = self._rbf_func(aug_particles.detach(),
+                                                    aug_particles.detach())
+        kernel_logp = torch.matmul(kernel_weight.t(),
```

Comment:
why transpose kernel_weight? It's symmetric.

Reply:
This was copied from the more general case when two entries to rbf_func might be different. Just corrected.